File: Model/cmyk.cpp
#include "cmyk.h"
#include <QString>
#include <math.h>
#include <typeinfo>
#include <iostream>
Cmyk::Cmyk(unsigned int c, unsigned int m, unsigned int y, unsigned int b) {
    if(c < CMYK_MAX_VALUE) {
        cyan=c;
    }
    else
        cyan = CMYK_MAX_VALUE;

    if(m < CMYK_MAX_VALUE){
        magenta = m;
    }
    else
        magenta = CMYK_MAX_VALUE;

    if(y < CMYK_MAX_VALUE){
        yellow = y;
    }
    else
        yellow = CMYK_MAX_VALUE;

    if(b < CMYK_MAX_VALUE) {
        black = b;
    }
    else
        black = CMYK_MAX_VALUE;

}

Cmyk::Cmyk(QString s) {
    QString r = s.mid(1,2);
    QString g = s.mid(3,2);
    QString b = s.mid(5,2);
    double r1=r.toInt(0,16);
    double r2=g.toInt(0,16);
    double r3=b.toInt(0,16);
    r1 = r1/255;
    r2 = r2/255;
    r3 = r3/255;
    double max = std::max(r1,r2);
    max = std::max(max,r3);
    double bl = 1-max;
    black = bl*100;
    double cy = (1-r1-bl) / (1-bl);
    cyan = cy*100;
    double mag = (1-r2-bl) / (1-bl);
    magenta = mag*100;
    double yel = (1-r3-bl) / (1-bl);
    yellow = yel * 100;

}

unsigned int Cmyk::getCyan() const {
    return cyan;
}

unsigned int Cmyk::getMagenta() const {
    return magenta;
}

unsigned int Cmyk::getYellow() const {
    return yellow;
}

unsigned int Cmyk::getBlack() const {
    return black;
}

void Cmyk::setCyan(unsigned int c) {
    if(c < CMYK_MAX_VALUE) {
        cyan=c;
    }
    else
        cyan = CMYK_MAX_VALUE;
}

void Cmyk::setMagenta(unsigned int m) {
    if(m < CMYK_MAX_VALUE){
        magenta = m;
    }
    else
        magenta = CMYK_MAX_VALUE;
}

void Cmyk::setYellow(unsigned int y) {
    if(y < CMYK_MAX_VALUE){
        yellow = y;
    }
    else
        yellow = CMYK_MAX_VALUE;
}

void Cmyk::setBlack(unsigned int b) {
    if(b < CMYK_MAX_VALUE) {
        black = b;
    }
    else
        black = CMYK_MAX_VALUE;
}

QString Cmyk::getHex() const {
    QString temp = "#";
    float c = cyan;
    float m = magenta;
    float y = yellow;
    float bl = black;

    int r = round(255*( 1 - c / 100) * ( 1 - bl / 100));
    int g = round(255*( 1 - m / 100) * ( 1 - bl / 100));
    int b = round(255*( 1 - y / 100) * ( 1 - bl / 100));

    temp += QString::number(r,16).rightJustified(2,'0');
    temp += QString::number(g,16).rightJustified(2,'0');
    temp += QString::number(b,16).rightJustified(2,'0');
    return temp;

}

unsigned int Cmyk::CMYK_MAX_VALUE = 100;

Cmyk* Cmyk::clone() const {
    return new Cmyk(*this);
}

Cmyk* Cmyk::operator+(const Colore& c) const {
    unsigned int c1 = cyan;
    unsigned int m = magenta;
    unsigned int y = yellow;
    unsigned int k = black;
    if(typeid(c)==typeid(const Cmyk&)) {
        const Cmyk* cmykp = static_cast<const Cmyk*>(&c);
        c1 += cmykp->getCyan();
        m += cmykp->getMagenta();
        y += cmykp->getYellow();
        k += cmykp->getBlack();
        return new Cmyk(c1,m,y,k);
    }
    else {
        Cmyk t = Cmyk(c.getHex());
        c1 += t.getCyan();
        m += t.getMagenta();
        y += t.getYellow();
        k += t.getBlack();
        return new Cmyk(c1,m,y,k);

    }

}

Cmyk* Cmyk::operator-(const Colore& c) const {
    int c1 = cyan;
    int m = magenta;
    int y = yellow;
    int k = black;
    if(typeid(c)==typeid(const Cmyk&)) {
        const Cmyk* cmykp = static_cast<const Cmyk*>(&c);
        c1 -= cmykp->getCyan();
        m -= cmykp->getMagenta();
        y -= cmykp->getYellow();
        k -= cmykp->getBlack();
        if(c1<0)
            c1=0;
        if(m<0)
            m=0;
        if(y<0)
            y=0;
        if(k<0)
            k=0;
        return new Cmyk(c1,m,y,k);
    }
    else {
        Cmyk t = Cmyk(c.getHex());
        c1 -= t.getCyan();
        m -= t.getMagenta();
        y -= t.getYellow();
        k -= t.getBlack();
        if(c1<0)
            c1=0;
        if(m<0)
            m=0;
        if(y<0)
            y=0;
        if(k<0)
            k=0;
        return new Cmyk(c1,m,y,k);

    }

}

Cmyk* Cmyk::media (const Colore& c) const {

    unsigned int c1 = cyan;
    unsigned int m = magenta;
    unsigned int y = yellow;
    unsigned int k = black;
    if(typeid(c)==typeid(const Cmyk&)) {
        const Cmyk* cmykp = static_cast<const Cmyk*>(&c);
        c1 += cmykp->getCyan();
        m += cmykp->getMagenta();
        y += cmykp->getYellow();
        k += cmykp->getBlack();
        return new Cmyk(abs(c1/2),abs(m/2),abs(y/2),abs(k/2));
    }
    else {
        Cmyk t = Cmyk(c.getHex());
        c1 += t.getCyan();
        m += t.getMagenta();
        y += t.getYellow();
        k += t.getBlack();
        return new Cmyk(abs(c1/2),abs(m/2),abs(y/2),abs(k/2));

    }
}
unsigned int Cmyk::getMaxValues() const {
    return CMYK_MAX_VALUE;
}

/*
Red = 255 × ( 1 - Cyan ÷ 100 ) × ( 1 - Black ÷ 100 )
Green = 255 × ( 1 - Magenta ÷ 100 ) × ( 1 - Black ÷ 100 )
Blue = 255 × ( 1 - Yellow ÷ 100 ) × ( 1 - Black ÷ 100 )
*/

File: SuperQtKalkJAVA/src/Cmyk.java

import static java.lang.Math.round;
import static java.lang.Math.abs;
import static java.lang.Math.max;

public class Cmyk implements Colore {
    private int cyan;
    private int magenta;
    private int yellow;
    private int black;
    private static int CMYK_MAX_VALUE = 100;

    public Cmyk(int c, int m, int y, int k) {
    	if(c > CMYK_MAX_VALUE)
    		cyan = CMYK_MAX_VALUE;
    	else if (c < 0)
    		cyan = 0;
    	else
    		cyan = c;
    	
    	if(m > CMYK_MAX_VALUE)
    		magenta = CMYK_MAX_VALUE;
    	else if (m < 0)
    		magenta = 0;
    	else
    		magenta = m;
    	
    	if(y > CMYK_MAX_VALUE)
    		yellow = CMYK_MAX_VALUE;
    	else if (y < 0)
    		yellow = 0;
    	else
    		yellow = y;
    	
    	if(k > CMYK_MAX_VALUE)
    		black = CMYK_MAX_VALUE;
    	else if (k < 0)
    		black = 0;
    	else
    		black = k;
    }

    public Cmyk(String s) {
        String r = s.substring(1,2);
        String g = s.substring(3,4);
        String b = s.substring(5,6);
        double r1 =Integer.parseInt(r,16);
        double g1 = Integer.parseInt(g,16);
        double b1 = Integer.parseInt(b,16);
        r1 = r1/255;
        g1 = g1/255;
        b1 = b1/255;
        double max = max(r1,g1);
        double max2=0;
        max2 = max(max,b1);
        double bla = 1 - max2;
        double cy = (1-r1-bla) / (1-bla);
        double mag = (1-g1-bla) / (1-bla);
        double yel = (1-b1-bla) / (1-bla);
        cyan = (int)round(cy*100);
        magenta = (int)round(mag*100);
        yellow = (int)round(yel*100);
        black = (int)round(bla*100);
    }

    public int getCyan() {
        return cyan;
    }
    public int getMagenta() {
        return magenta;
    }
    public int getYellow() {
        return yellow;
    }
    public int getBlack() {
        return black;
    }

    void setCyan(int c) {
    	if(c > CMYK_MAX_VALUE)
    		cyan = CMYK_MAX_VALUE;
    	else if (c < 0)
    		cyan = 0;
    	else
    		cyan = c;
    }

    void setMagenta(int m) {
    	if(m > CMYK_MAX_VALUE)
    		magenta = CMYK_MAX_VALUE;
    	else if (m < 0)
    		magenta = 0;
    	else
    		magenta = m;
    }

    void setYellow(int y) {
    	if(y > CMYK_MAX_VALUE)
    		yellow = CMYK_MAX_VALUE;
    	else if (y < 0)
    		yellow = 0;
    	else
    		yellow = y;
    }

    void setBlack(int k) {
    	if(k > CMYK_MAX_VALUE)
    		black = CMYK_MAX_VALUE;
    	else if (k < 0)
    		black = 0;
    	else
    		black = k;
    }

    
    public String getHex() {
        String temp = "#";
        double c = cyan;
        double m = magenta;
        double y = yellow;
        double bl = black;

        double r1 = round(255*( 1 - c / 100) * ( 1 - bl / 100));
        double g1 = round(255*( 1 - m / 100) * ( 1 - bl / 100));
        double b1 = round(255*( 1 - y / 100) * ( 1 - bl / 100));

        String r = String.format("%02X", (int)r1);
        String g = String.format("%02X", (int)g1);
        String b = String.format("%02X", (int)b1);
        return temp + r + g + b;
    }

    
    public Cmyk somma(Colore c) {
        int c1 = cyan;
        int m = magenta;
        int y = yellow;
        int k = black;
        if(c instanceof Cmyk) {
            Cmyk temp = (Cmyk) c;
            c1 += temp.getCyan();
            m += temp.getMagenta();
            y += temp.getYellow();
            k += temp.getBlack();
        }
        else {
            Cmyk t = new Cmyk(c.getHex());
            c1 += t.getCyan();
            m += t.getMagenta();
            y += t.getYellow();
            k += t.getBlack();

        }
        return new Cmyk(c1,m,y,k);

    }

    
    public Cmyk sottrazione(Colore c) {
        int c1 = cyan;
        int m = magenta;
        int y = yellow;
        int k = black;
        if(c instanceof Cmyk) {
            Cmyk temp = (Cmyk) c;
            c1 -= temp.getCyan();
            m -= temp.getMagenta();
            y -= temp.getYellow();
            k -= temp.getBlack();
        }
        else {
            Cmyk t = new Cmyk(c.getHex());
            c1 -= t.getCyan();
            m -= t.getMagenta();
            y -= t.getYellow();
            k -= t.getBlack();

        }
        if(c1<0) c1 = 0;
        if(m<0) m = 0;
        if(y<0) y = 0;
        if(k<0) k = 0;
        return new Cmyk(c1,m,y,k);

    }

    
    public Cmyk media(Colore c) {
        int c1 = cyan;
        int m = magenta;
        int y = yellow;
        int k = black;
        if(c instanceof Cmyk) {
            Cmyk temp = (Cmyk) c;
            c1 += temp.getCyan();
            m += temp.getMagenta();
            y += temp.getYellow();
            k += temp.getBlack();
        }
        else {
            Cmyk t = new Cmyk(c.getHex());
            c1 += t.getCyan();
            m += t.getMagenta();
            y += t.getYellow();
            k += t.getBlack();

        }
        return new Cmyk(abs(c1/2),abs(m/2),abs(y/2),abs(k/2));

    }
    
    public int getMaxValues() {
        return CMYK_MAX_VALUE;
    }
}
